src/raftstore/store/snap.rs

```
@@ -618,6 +620,987 @@ mod v1 {
+            let mut checksum_value = Vec::with_capacity(number::U64_SIZE);
+            checksum_value.encode_u64(cf_file.checksum as u64).unwrap();
+            let mut kv = KeyValue::new();
+            kv.set_key(checksum_key_buf);
+            kv.set_value(checksum_value);
+            kvs.push(kv);
+        }
+        Ok((total_size, kvs))
+    }
+
+
+    fn decode_cf_size_and_checksums(kvs: &[KeyValue]) -> RaftStoreResult<Vec<(CfName, u64, u32)>> {
+        let snapshot_cfs = get_snapshot_cfs();
+        let mut cf_sizes: Vec<(CfName, u64)> = vec![];
+        let mut cf_checksums: Vec<(CfName, u32)> = vec![];
+        for kv in kvs {
+            let mut key_bytes = kv.get_key();
+            let decoded = try!(key_bytes.decode_bytes(ENCODE_DECODE_DESC));
+            let key = match String::from_utf8(decoded) {
+                Ok(s) => s,
+                Err(e) => {
+                    return Err(box_err!("fail to parse snapshot meta key {:?}, err: {:?}",
+                                        key_bytes,
+                                        e));
+                }
+            };
+            let strs: Vec<&str> = key.split('_').collect();
+            if strs.len() != 2 {
+                return Err(box_err!("invalid snapshot meta key {}", key));
+            }
+            let cf_name = strs[1];
+            let res = snapshot_cfs.iter().find(|&s| *s == cf_name);
+            if res.is_none() {
+                return Err(box_err!("invalid snapshot cf {}", cf_name));
+            }
+            let cf = res.unwrap();
+            match strs[0] {
+                SNAPSHOT_META_PREFIX_SIZE => {
+                    let size = try!(kv.get_value().decode_u64());
+                    cf_sizes.push((cf, size));
+                }
+                SNAPSHOT_META_PREFIX_CHECKSUM => {
+                    let value = try!(kv.get_value().decode_u64());
+                    if value > u32::max_value() as u64 {
```

You can check the checksum here to fail fast.